### DISCUSSION TITLE
undescore in link name is escaped

### DISCUSSION THREAD

**Discussion by muravjev** [remarkjs/remark#1067] [Q&A] [answered]
## Prerequisites

1. We have a link that contains a name with underscore in it:

```md
[CODE_OF_CONDUCT.md]()
```

2. After running remark:

```sh
remark --output -- .
```

3. Link name is formatted:

```md
[CODE\_OF\_CONDUCT.md]()
```

diff: https://github.com/muravjev/repro_remark_underscore/compare/0d32000...defd596

## Question

Is it possible to prevent such behavior?

## Reproduction

repo: https://github.com/muravjev/repro_remark_underscore

```sh
git clone https://github.com/muravjev/repro_remark_underscore.git
cd repro_remark_underscore
pnpm install
pnpm repro
```

**Comment by ChristianMurphy** [comment, marked as answer]
`remark` is an [abstract syntax tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) tool.
Meaning it aims to preserve structure/meaning of content, but may not use the exact same characters.
Put another way it acts as a formatter.

The examples you provide are an example of this:

```markdown
[CODE_OF_CONDUCT.md]()
[CODE\_OF\_CONDUCT.md]()
```

render exactly the same:

---

[CODE_OF_CONDUCT.md]()
[CODE\_OF\_CONDUCT.md]()

---

the escapes are added to make it clear that the content is not emphasized.

This behavior is not configurable, and the aim of exactly preserving characters doesn't really fit in the remark ecosystem.
To robustly preserve the exact characters/tokens, a concrete syntax tree, a related but different data structure would be needed.
That would be its own project/ecosystem building on top of the [micromark](https://github.com/micromark/micromark#readme) tokenizer.

---

Taking a step back, this does sound a bit like an [XY problem](https://xyproblem.info/).
What (if any) problem are the escapes creating?

Is another tool having trouble parsing them?
If so, that would be something to raise with that project, as it diverges from the [commonmark standard](https://commonmark.org), which much of the markdown community aims to adhere to.

**Reply by muravjev** [reply]
OK

But for me it looks like a defect. I work with markdown source, not output, and for me `CODE_OF_CONDUCT` is much more readable and comfortable to edit, navigate etc., then  `CODE\_OF\_CONDUCT`.

Its like you write `const i=4;` and some linter(formatter) will change it to `const i=1+1+5-3;`, yeah, the code will produce _the same_ output, but ...
